### PR TITLE
Explicitly track relative age in the ResourceRegistry

### LINKED ISSUE #2897: Nested allocation leads to out-of-order ResourceRegistry keys
If the argument to `alloc` itself also allocates something in the registry, then the keys will no longer accurately reflect the relative age of the objects. This means the registry will deallocate them in the wrong order.

Optional task component: auditing the code base to see if this use pattern currently occurs.

We set medium priority because we haven't yet identified concrete misbehavior from this: we just identified a shortcoming of the internal library we're relying on to ensure that our cleanup handlers get to run correctly, which is a subtle thing. (EG releasing a lock when we're shutting down the DB, sending a goodbye message on a socket, etc.)

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes #2897.

### CODE REVIEW

**Comment by nfrisby** [comment]
@edsko will review this live during Thursday's Meet. @EncodePanda, I'd like you to give it a pass too. It's self-contained and the module is very well documented, so you can look as deep as you like in your review pass.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 1]
```diff
@@ -307,13 +325,28 @@ data RegistryState m = RegistryState {
       -- | Next available resource key
     , registryNextKey   :: !ResourceId
 
+      -- | The age of each currently allocated resource
+      --
+      -- We use a 'Bimap' so we can both efficiently forget a resource when it
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:330]: Simplify this; just mention the ordering, let the forgetting part be implicit.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 2]
```diff
@@ -296,6 +300,20 @@ deriving instance IOLike m => NoThunks (ResourceRegistry m)
   Internal: registry state
 -------------------------------------------------------------------------------}
 
+-- | The age of a resource is represented by how many resources were allocated
+-- before it
+--
+-- This data is not exposed by the 'ResourceRegistry' interface.
+newtype Age = YoungerThan Word64
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:307]: Edsko favors the `maxBound` idea, let's try that.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 3]
```diff
@@ -411,14 +442,26 @@ insertResource :: ResourceId
...
+      , registryAges      = Bimap.insert
+                              key
+                              (registryNextAge st)
+                              (registryAges st)
+      , registryNextAge   = nextYoungerAge (registryNextAge st)
       }
 
 -- | Remove resource from the registry (if it exists)
 removeResource :: ResourceId -> State (RegistryState m) (Maybe (Resource m))
-removeResource key = state $ \st ->
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:418]: Try to use `state`, but with less point-free.

**Review by edsko — APPROVED** [review]
Nice, good to decouple these two concepts.

**Pull request merged**